src/include/daos/rsvc.h

```diff
@@ -51,6 +51,9 @@ struct rsvc_client {
 	uint64_t	sc_leader_term;
 	int		sc_leader_index;	/* in sc_ranks */
 	int		sc_next;		/* in sc_ranks */
+	uuid_t		sc_id;			/* of rsvc e.g., uuid */
+	void	       *sc_priv;		/* caller-private info */
+	int	      (*sc_update_ranks_cb)(struct rsvc_client *);
```

Comment:
Have you considered querying MS in dc_pool and dc_cont, instead of changing rsvc_client? For example, dc_pool could call its own "choose" function, which handles -DER_NOTREPLICAs from rsvc_client_choose by querying the MS:

    some_dc_pool_choose_function()
    {
        rc = rsvc_client_choose(...);
        if (rc == -DER_NOTREPLICA) {
            dc_mgmt_get_pool_svc_ranks(...);
            rsvc_client_fini(...);
            rsvc_client_init(...);
        }
    }

That would keep the locking (i.e., querying while holding dc_pool lock), blocking (i.e., blocking the DAOS API caller), rsvc ID type (i.e., uuid_t versus d_iov_t) questions out of dc_rsvc.